### ISSUE TITLE
Implement positional tracking without cockpit shake

### ISSUE THREAD

**Issue by morallo** [Prof-Butts/Hook_XWACockpitLook#9]
Using the cockpit shake variables seems to introduce a lot of jitter.

If the actual camera position parameters are used instead, they get overwritten later in the frame processing pipeline.
I think we need to identify where this is happening and preventing it with a specific hook.

**Comment by morallo** [comment]
This is the `PlayerCamera` struct from IDA:
```
/* 799 */
struct __unaligned __declspec(align(1)) PlayerCamera
{
  int PositionX;
  int PositionY;
  int PositionZ;
  int CraftIndex;
  int RelatedToMap;
  __int16 CraftPitch;
  __int16 CraftYaw;
  __int16 CraftRoll;
  __int16 Roll;
  __int16 Pitch;
  __int16 Yaw;
  int ShakeX;
  int ShakeY;
  int ShakeZ;
  __int16 collisionPitch;
  __int16 collisionYaw;
  __int16 collisionRoll;
  __int16 field_32;
  char ViewMode1;
  char ViewMode2;
  char field_36;
  char S0x08B94E0_B48_m37;
  char unk38;
  __int16 S0x08B94E0_B48_m39;
  __int16 S0x08B94E0_B48_m3B;
  __int16 MapMode;
  __int16 S0x08B94E0_B48_m3F;
  __int16 ExternalCamera;
  int ExternalCameraZoomDist;
  __int16 S0x08B94E0_B48_m47;
  __int16 S0x08B94E0_B48_m49;
  char unk4B[2];
};
```

Currently the values set by CockpitLook are ShakeX, ShakeY, ShakeZ.
We should modify PositionX, PositionY, PositionZ instead.

**Comment by Prof-Butts** [comment]
This sounds easy to test. Have you checked what happens if you use PositionX/Y/Z instead?

**Comment by morallo** [comment]
Yes, it does not work. I guess it gets overwritten later in the frame processing. I'm trying to locate where with IDA.

**Comment by morallo** [comment]
Currently the CockpitLook hooks the start of `PlayerInflightSystemInput` at `0x4F9A66`.
There is a function tagged as `PlayerCameraUpdate` at `0x4EE820` that may be a better candidate.
I will do some tests.

**Comment by morallo** [comment]
I still didn't figure out where to hook, PlayerCameraUpdate seems to do a lot of stuff.

However I found this inside it:
```
    PlayerDataTable[v4].Camera.PositionX += PlayerDataTable[v4].Camera.ShakeX >> 4;
    PlayerDataTable[v4].Camera.PositionY += PlayerDataTable[v4].Camera.ShakeY >> 4;
    PlayerDataTable[v4].Camera.PositionZ += PlayerDataTable[v4].Camera.ShakeZ >> 4;
```
If I understand correctly, it is doing a 4-bit right shift to the shake values. Could it be the source of the jerkiness in the positional tracking when using those values? Even if the total movement is scaled somewhere else, whith the shift we lose 4 bits of precision.

**Comment by morallo** [comment]
> 
> ```
>     PlayerDataTable[v4].Camera.PositionX += PlayerDataTable[v4].Camera.ShakeX >> 4;
>     PlayerDataTable[v4].Camera.PositionY += PlayerDataTable[v4].Camera.ShakeY >> 4;
>     PlayerDataTable[v4].Camera.PositionZ += PlayerDataTable[v4].Camera.ShakeZ >> 4;
> ```
> 
> If I understand correctly, it is doing a 4-bit right shift to the shake values. Could it be the source of the jerkiness in the positional tracking when using those values? Even if the total movement is scaled somewhere else, whith the shift we lose 4 bits of precision.

I tried patching the _shift arithmetic right_ instruction with NOP's with the hook, but the positional tracking is still jerky.

It's not really jitter. I'm convinced it is a lack of precision. If you translate your head slowly, the view only changes in noticeable "steps".

**Comment by Prof-Butts** [comment]
Yes, we're losing 4 bits of precision here and yes, that's going to cause some jittering. Maybe the best way to proceed here would be to hook the function at `0x4EE820` to overwrite the values in `PlayerDataTable[v4].Camera.PositionX/Y/Z`.

I'll see if I can write a simple hook (unless you want to give it a try). One way to start writing a hook is by taking the CockpitLookHook and reduce it to a no-op. Then we can hook it to the right address to start modifying the `PlayerDataTable`

**Comment by morallo** [comment]
> I'll see if I can write a simple hook (unless you want to give it a try). One way to start writing a hook is by taking the CockpitLookHook and reduce it to a no-op. Then we can hook it to the right address to start modifying the `PlayerDataTable`

I did the NOPs already by adding some patches in CockpitLook's `hooks.h`, but that did not fix the jerkiness.

I don't think there is a need for a new hook. We can hook different places in the same CockpitLook and call different functions.

I found this near the end of the `PlayerCameraUpdate` (0x4EE820)
```
    PlayerDataTable[playerIndex].Camera.PositionX = g_objectPtr[PlayerDataTable[playerIndex].Camera.CraftIndex].PosX;
    PlayerDataTable[playerIndex].Camera.PositionY = objPtr[PlayerDataTable[playerIndex].Camera.CraftIndex].PosY;
    PlayerDataTable[playerIndex].Camera.PositionZ = objPtr[PlayerDataTable[playerIndex].Camera.CraftIndex].PosZ;
    if ( PlayerDataTable[playerIndex].Camera.CraftIndex == PlayerDataTable[playerIndex].ObjectIndex )
    {
        UpdatePlayerMovement(playerIndexCount);
        PlayerDataTable[playerIndexCount].Camera.PositionX += (signed __int64)PlayerDataTable[playerIndexCount].CockpitPositionTranformed.x;
        PlayerDataTable[playerIndexCount].Camera.PositionY += (signed __int64)PlayerDataTable[playerIndexCount].CockpitPositionTranformed.y;
        v16 = (signed __int64)PlayerDataTable[playerIndexCount].CockpitPositionTranformed.z;
        objPtr = g_objectPtr;
        PlayerDataTable[playerIndex].Camera.PositionZ += v16;
    }
    PlayerDataTable[playerIndex].Camera.PositionX += PlayerDataTable[playerIndex].Camera.ShakeX >> 4;
    PlayerDataTable[playerIndex].Camera.PositionY += PlayerDataTable[playerIndex].Camera.ShakeY >> 4;
    PlayerDataTable[playerIndex].Camera.PositionZ += PlayerDataTable[playerIndex].Camera.ShakeZ >> 4;
```
If I understand that correctly, it's doing this:
- Reset the position of the camera at the position of the craft in the world.
- Add offset between the origin of the model and the cockpit position.
- Apply shake.

That would explain why setting the PositionX,Y,Z does not work at the currently hooked function (it happens at the "capture player input" phase of the frame, when the mouse position is taken to set the pitch+yaw).

I am going to try to hook that and run a function that adds the headtracking offset.

**Comment by morallo** [comment]
The new hook works at that place, but only using the same Shake variables.
If I NOP the modification of Camera.PositionX,Y,Z with the Shake at PlayerUpdateCamera, the tracking still works, so even if the values are set at that function by the hook, they are actually used later to modify the camera.

Debugging I verified that these locations access the memory address of ShakeX each frame:
- 0x004959D8 inside UpdateObjectMovement()
- 0x004EEA89 inside PlayerCameraUpdate() --> this is the one decompiled in the previous comment.
- **0x00480B50 inside an unknown function sub_480A80**.
- ddraw.PrimarySurface::ComputeAddGeomViewMatrix

This is the relevant part of the code in sub_480A80 after some variable renaming:
```
    if ( (_WORD)mobileObject )
    {
        v2 = (OPTHeader *)Lock_Handle((__int16)mobileObject);
        PtrAdjustOptimizedPolyObject(v2);
        memset(&sceneCompData, 0, sizeof(sceneCompData));
        objectPosX = param1_object->PosX;
        sceneCompData.pObject = param1_object;
        positionX_offset = objectPosX - PlayerDataTable[g_currentPlayerIndex].Camera.PositionX;
        objectPosZ = param1_object->PosZ;
        offset_x_double = (double)positionX_offset;
        positionY_offset = param1_object->PosY - PlayerDataTable[g_currentPlayerIndex].Camera.PositionY;
        cameraPositionZ = PlayerDataTable[g_currentPlayerIndex].Camera.PositionZ;
        *(float *)&sceneCompData.ViewTransform.Position.positionX = offset_x_double;
        *(float *)&sceneCompData.ViewTransform.Position.positionY = (float)positionY_offset;
        *(float *)&sceneCompData.ViewTransform.Position.positionZ = (float)(objectPosZ - cameraPositionZ);
        if ( dword_91AE7C )
        {
            shakeY = (double)PlayerDataTable[g_currentPlayerIndex].Camera.ShakeY;
            shakeZ = (double)PlayerDataTable[g_currentPlayerIndex].Camera.ShakeZ;
            cockpitPositionTransformedZ = PlayerDataTable[g_currentPlayerIndex].CockpitPositionTranformed.z;
            cockpitPositionTransformedY = -PlayerDataTable[g_currentPlayerIndex].CockpitPositionTranformed.y;
            *(float *)&sceneCompData.ViewTransform.Position.positionX = -PlayerDataTable[g_currentPlayerIndex].CockpitPositionTranformed.x
                                                                      - (double)PlayerDataTable[g_currentPlayerIndex].Camera.ShakeX
                                                                      * 0.0625;
            *(float *)&sceneCompData.ViewTransform.Position.positionY = cockpitPositionTransformedY - shakeY * 0.0625;
            *(float *)&sceneCompData.ViewTransform.Position.positionZ = -cockpitPositionTransformedZ - shakeZ * 0.0625;
        }
        sceneCompData.ViewTransform.Rotation._11 = flt_7B4BEC;
        sceneCompData.ViewTransform.Rotation._12 = flt_7B4BE8;
        sceneCompData.ViewTransform.Rotation._21 = flt_7B6FF8;
        sceneCompData.ViewTransform.Rotation._22 = flt_7B6FF0;
        sceneCompData.ViewTransform.Rotation._13 = flt_7B4BF4;
        sceneCompData.ViewTransform.Rotation._31 = flt_7B33DC;
        sceneCompData.ViewTransform.Rotation._32 = flt_7B33D8;
        sceneCompData.ViewTransform.Rotation._23 = flt_7B33D4;
        sceneCompData.ViewTransform.Rotation._33 = flt_7B4BE4;
        Vector3Transform((Vector3_float *)&sceneCompData.ViewTransform, &sceneCompData.ViewTransform.Rotation);
        sceneCompData.ViewTransform.Rotation._11 = flt_7C9DA8;
        sceneCompData.ViewTransform.Rotation._12 = flt_7C9DAC;
        sceneCompData.ViewTransform.Rotation._21 = flt_7CA15C;
        sceneCompData.ViewTransform.Rotation._22 = flt_7CA160;
        sceneCompData.ViewTransform.Rotation._13 = flt_7C9DB4;
        sceneCompData.ViewTransform.Rotation._31 = flt_7CA168;
        sceneCompData.ViewTransform.Rotation._32 = flt_7CA170;
        sceneCompData.ViewTransform.Rotation._23 = flt_7CA164;
        sceneCompData.WorldTransform.Rotation._11 = flt_7C9DA8;
        sceneCompData.WorldTransform.Rotation._12 = flt_7CA15C;
        sceneCompData.ViewTransform.Rotation._33 = flt_7CA178;
        sceneCompData.WorldTransform.Rotation._21 = flt_7C9DAC;
        sceneCompData.WorldTransform.Rotation._22 = flt_7CA160;
        *(float *)&sceneCompData.WorldTransform.Position.positionX = -*(float *)&sceneCompData.ViewTransform.Position.positionX;
        sceneCompData.WorldTransform.Rotation._13 = flt_7CA168;
        *(float *)&sceneCompData.WorldTransform.Position.positionY = -*(float *)&sceneCompData.ViewTransform.Position.positionY;
        sceneCompData.WorldTransform.Rotation._31 = flt_7C9DB4;
        sceneCompData.WorldTransform.Rotation._32 = flt_7CA164;
        sceneCompData.WorldTransform.Rotation._23 = flt_7CA170;
        *(float *)&sceneCompData.WorldTransform.Position.positionZ = -*(float *)&sceneCompData.ViewTransform.Position.positionZ;
        sceneCompData.WorldTransform.Rotation._33 = flt_7CA178;
        Vector3Transform((Vector3_float *)&sceneCompData.WorldTransform, &sceneCompData.WorldTransform.Rotation);
```

It looks promising! I will try to hook directly the camera position there.

**Comment by morallo** [comment]
Fixed in #10 and #14
